FilteringTargetWrapper: Add support for batch writing + PostFilteringTargetWrapper: performance optimizations

snakefoot opened the pull request:
And allow use of WhenRepeated as Filter:

```xml
      <target name="default" xsi:type="FilteringWrapper">
        <filter xsi:type="whenRepeated" layout="${message}" timeoutSeconds="30" action="Ignore" />
        <target name="console" xsi:type="Console" />
      </target>
```

Also made some optimizations to `PostFilteringTargetWrapper` so it also supports OptimizeBufferReuse = true.

And also "fixed" `PostFilteringTargetWrapper` so it also works when writing a single LogEvent, without using BufferingWrapper or AsyncWrapper (More intuitive and user-friendly)

304NotModified commented:
Thanks!

Is this a Breaking change? (Removal of ConditionExpression Condition)

snakefoot commented:
> Is this a Breaking change? (Removal of ConditionExpression Condition)

Nope this property:

> public ConditionExpression Condition { get; set; }

Is now forwarding into the new `Filter`-property using `ConditionBasedFilter`

snakefoot commented:
@304NotModified Please include for ver. 4.6.4

snakefoot commented on src/NLog/Internal/ArrayHelper.cs:
```diff
@@ -31,19 +31,34 @@
+        /// <param name="untilIndex">until index, not included.</param>
+        /// <returns>New list</returns>
+        /// <remarks>IList as input for performance reasons.</remarks>
+        public static IList<T> CreatePartialList<T>([NotNull] this IList<T> items, int untilIndex)
```
Maybe call this `Take` or `TakeList` ?

304NotModified commented:
It would be nice if we could deduplicate PostFilteringTargetWrapper.ApplyFilter and FilteringTargetWrapper.Write - any idea how?

304NotModified commented on src/NLog/Targets/Wrappers/FilteringTargetWrapper.cs:
```diff
@@ -90,16 +89,32 @@ public FilteringTargetWrapper(Target wrappedTarget, ConditionExpression conditio
         /// to the wrapped target.
         /// </summary>
         /// <docgen category='Filtering Options' order='10' />
+        public ConditionExpression Condition { get => (Filter as ConditionBasedFilter)?.Condition; set => Filter = CreateFilter(value); }
```
FYI I created a helper here, so the line would be more readable.

304NotModified commented on src/NLog/Config/LoggingConfigurationParser.cs:
```diff
@@ -1095,6 +1100,30 @@ private Layout TryCreateLayoutInstance(ILoggingConfigurationElement element, Typ
             return _configurationItemFactory.Layouts.CreateInstance(ExpandSimpleVariables(layoutTypeName));
         }
 
+        private bool SetFilterFromElement(object o, PropertyInfo propInfo, ILoggingConfigurationElement layoutElement)
```
Please don't copy paste code.

I've added an abstraction method for this.

snakefoot commented:
> It would be nice if we could deduplicate PostFilteringTargetWrapper.ApplyFilter and FilteringTargetWrapper.Write - any idea how?

Made a commit

304NotModified commented:
cool,

what do you think of https://github.com/snakefoot/NLog/pull/8 ? Or are lambdas bad for performance?

snakefoot commented:
> Or are lambdas bad for performance?

Yes. Those that are doing capture. Was also doing it first. Until the other target comes with a dynamic-filter-object. Then it became a TState-parameter.

304NotModified commented:
OK! Added some docs and now this is perfect :)

snakefoot commented:
> OK! Added some docs and now this is perfect IMO :)

Also happy with the result :)

snakefoot commented:
@304NotModified Guess you need to restart the appveyor-build

304NotModified commented:
restarted 2nd time 👼

304NotModified commented:
Oops the idea was to merge this on to release/4.6.4. Anyway, canceled the build so no problem.

snakefoot commented:
Updated the documentation: https://github.com/NLog/NLog/wiki/FilteringWrapper-target and added link from https://github.com/NLog/NLog/wiki/WhenRepeated-Filter

304NotModified commented:
Ah this is also an xsd update
